java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java

```diff
@@ -0,0 +1,237 @@
+
+  @Override
+  public void init(MethodTree tree) {
+    methodTree = tree;
+  }
+
+  private static final String[] ANNOTATIONS = {"javax.annotation.Nonnull", "javax.validation.constraints.NotNull",
+    "edu.umd.cs.findbugs.annotations.NonNull", "org.jetbrains.annotations.NotNull", "lombok.NonNull",
+    "android.support.annotation.NonNull"};
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    final AbstractStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    final PostStatementVisitor visitor = new PostStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  @Override
+  public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager constraintManager) {
+    if (methodTree.is(Tree.Kind.CONSTRUCTOR)) {
+      ClassTree classTree = (ClassTree) methodTree.parent();
+      for (Tree member : classTree.members()) {
+        if (member.is(Tree.Kind.VARIABLE)) {
+          checkVariable(context, methodTree, (VariableTree) member);
+        }
+      }
+    }
+  }
+
+  private void checkVariable(CheckerContext context, MethodTree tree, VariableTree member) {
+    String nonNullAnnotation = nonNullAnnotation((JavaSymbol) member.symbol());
+    if (nonNullAnnotation != null && isUndefinedOrNull(context, member.symbol())) {
+      context.reportIssue(tree, this,
+        MessageFormat.format("\"{0}\" is marked \"{1}\" but is not initialized in this constructor.", member.symbol().name(), nonNullAnnotation));
+    }
+  }
+
+  private static boolean isUndefinedOrNull(CheckerContext context, Symbol symbol) {
+    ProgramState programState = context.getState();
+    SymbolicValue value = programState.getValue(symbol);
+    return value == null;
+  }
+
+  protected String nonNullAnnotation(JavaSymbol javaSymbol) {
+    for (String annotation : ANNOTATIONS) {
+      if (javaSymbol.metadata().isAnnotatedWith(annotation)) {
+        return annotation;
+      }
+    }
+    return null;
+  }
+
+  private class AbstractStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private final CheckerContext context;
+
+    protected AbstractStatementVisitor(CheckerContext context) {
+      super(context.getState());
+      this.context = context;
+    }
+
+    protected void reportIssue(Tree tree, String message, Object... parameters) {
+      context.reportIssue(tree, NonNullSetToNullCheck.this, MessageFormat.format(message, parameters));
+    }
+  }
+
+  private class PreStatementVisitor extends AbstractStatementVisitor {
+
+    protected PreStatementVisitor(CheckerContext context) {
+      super(context);
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      if (tree.variable().is(Tree.Kind.IDENTIFIER)) {
```

Comment:
At least one other case than assignment to a variable should be covered by unit tests.